### PR TITLE
Add Visual Code catalog generator

### PR DESCRIPTION
**Pull request by manuelcanepa**
### Description (*)
Add urn-catalog model generate for visual code

### Manual testing scenarios (*)
1. Run `bin/magento dev:urn-catalog:generate --ide=vscode  .vscode/catalog.xml`
2. Open project on visual code with redhat.vscode-xml extension

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+     * @param DomDocumentFactory $domDocumentFactory
+     */
+    public function __construct(
+        ReadFactory $readFactory,
+        WriteFactory $fileWriteFactory,
+        DomDocumentFactory $domDocumentFactory = null
+    ) {
+        $this->currentDirRead = $readFactory->create(getcwd());
+        $this->fileWriteFactory = $fileWriteFactory;
+        $this->domDocumentFactory = $domDocumentFactory ?: ObjectManager::getInstance()->get(DomDocumentFactory::class);
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php]: When creating a completely new class - you don't need to provide backward compatible class loader.

File: app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
+<?php
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php]: When introducing new class - use `strict_types=1`

File: app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    /**
+     * Generate Catalog of URNs for the VsCode
+     *
+     * @param string[] $dictionary
+     * @param string $configFilePath relative path to the PhpStorm misc.xml
+     * @return void
+     */
+    public function generateCatalog(array $dictionary, $configFilePath)
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php]: When introducing new class - use parameter type hints:
```php
    public function generateCatalog(array $dictionary, string $configFilePath): void
```

File: app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    {
+        $catalogNode = null;
+
+        try {
+            $file = $this->fileWriteFactory->create(
+                $configFilePath,
+                \Magento\Framework\Filesystem\DriverPool::FILE,
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php]: Import class, then it should fit 1 line.

File: app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                'r'
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php]: Replace literal with const - for example `FILE_MODE_READ`.

File: app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            $node = $dom->createElement('system');
+            $node->setAttribute('systemId', $urn);
+            $node->setAttribute('uri', $this->getFileLocationInProject($xsdPath));
+            $catalogNode->appendChild($node);
+        }
+        $dom->formatOutput = true;
+        $file = $this->fileWriteFactory->create(
+            $configFilePath,
+            \Magento\Framework\Filesystem\DriverPool::FILE,
+            'w'
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php]: const `FILE_MODE_WRITE`

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]

**Comment by lbajsarowicz** [comment]
Please provide Integration Tests to cover the feature.

File: app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php (hunk 3 above)
Comment (manuelcanepa) [reply app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php]: I cant
`Declaration of Magento\Developer\Model\XmlCatalog\Format\VsCode::generateCatalog(array $dictionary, string $configFile): void must be compatible with Magento\Developer\Model\XmlCatalog\Format\FormatInterface::generateCatalog(array $dictionary, $configFile)`

Comment (lbajsarowicz) [reply app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php]: Ok. I see. Just ignore this one.

**Review by lbajsarowicz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-6554](https://jira.corp.magento.com/browse/ENGCOM-6554) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by magento-engcom-team** [comment]
@manuelcanepa thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by engcom-Alfa** [comment]
Hi @manuelcanepa   .
     Due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all code must be covered by tests. Please cover your fix by auto test.
Thanks!

**Comment by manuelcanepa** [comment]
@engcom-Alfa A test like the phpStorm format will be enough?

**Comment by Nazar65** [comment]
@manuelcanepa i think is the best way is to cover `app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php`  by unit test.

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+use Magento\Framework\Filesystem\Directory\ReadInterface;
+use Magento\Framework\Filesystem\DriverPool;
+use Magento\Framework\Filesystem\File\WriteFactory;
+
+class VsCodeTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * @var Magento\Developer\Model\XmlCatalog\Format\VsCode
+     */
+    protected $vscodeFormat;
```
Comment (Nazar65) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: use private property

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+     * @var Magento\Framework\Filesystem\File\WriteFactory
+     */
+    protected $fileWriteFactory;
+
+    /**
+     * @var Magento\Framework\DomDocument\DomDocumentFactory
+     */
+    protected $domFactory;
+
+    protected $dictionary = [
```
Comment (Nazar65) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: you can use @dataProvider for that

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        $this->readFactory = $this->createMock(ReadFactory::class);
+        $this->readFactory->expects($this->once())
+            ->method('create')
+            ->withAnyParameters()
+            ->willReturn($currentDirRead);
+
+        $this->fileWriteFactory = $this->createMock(WriteFactory::class);
+        $this->domFactory = new DomDocumentFactory();
+
+        $this->vscodeFormat = new VsCode(
```
Comment (Nazar65) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: Use `ObjectManagerHelper` to Create instance of TestObject

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+            $this->readFactory,
+            $this->fileWriteFactory,
+            $this->domFactory
+        );
+    }
+
+    public function testGenerateNewValidCatalog()
```
Comment (Nazar65) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: Add a Doc Block for Methods

**Review by Nazar65 — CHANGES_REQUESTED** [review]

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+use Magento\Developer\Model\XmlCatalog\Format\VsCode;
+use Magento\Framework\DomDocument\DomDocumentFactory;
+use Magento\Framework\Exception\FileSystemException;
+use Magento\Framework\Filesystem\Directory\ReadFactory;
+use Magento\Framework\Filesystem\Directory\ReadInterface;
+use Magento\Framework\Filesystem\DriverPool;
+use Magento\Framework\Filesystem\File\WriteFactory;
+
+class VsCodeTest extends \PHPUnit\Framework\TestCase
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: Import class

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+{
+    /**
+     * @var Magento\Developer\Model\XmlCatalog\Format\VsCode
+     */
+    protected $vscodeFormat;
+
+    /**
+     * @var Magento\Framework\Filesystem\Directory\ReadFactory
+     */
+    protected $readFactory;
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: `readFactoryMock`

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+class VsCodeTest extends \PHPUnit\Framework\TestCase
+{
+    /**
+     * @var Magento\Developer\Model\XmlCatalog\Format\VsCode
+     */
+    protected $vscodeFormat;
+
+    /**
+     * @var Magento\Framework\Filesystem\Directory\ReadFactory
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: MockObject|ReadFactory

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+     */
+    protected $readFactory;
+
+    /**
+     * @var Magento\Framework\Filesystem\File\WriteFactory
+     */
+    protected $fileWriteFactory;
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: `fileWriteFactoryMock`

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+     */
+    protected $vscodeFormat;
+
+    /**
+     * @var Magento\Framework\Filesystem\Directory\ReadFactory
+     */
+    protected $readFactory;
+
+    /**
+     * @var Magento\Framework\Filesystem\File\WriteFactory
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: `MockObject|WriteFactory`

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+        $this->fileWriteFactory->expects($this->at(0))
+            ->method('create')
+            ->with(
+                $configFile,
+                DriverPool::FILE,
+                VsCode::FILE_MODE_READ
+            )
+            ->willThrowException(new FileSystemException($message));
+
+        $file = $this->createMock(\Magento\Framework\Filesystem\File\Write::class);
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: Import class.

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+        $this->vscodeFormat->generateCatalog($this->dictionary, $configFile);
+    }
+
+    public function testGenerateExistingValidCatalog()
+    {
+        $configFile = 'test';
+        $fixtureXmlFile = __DIR__ . '/_files/valid_catalog.xml';
+        $content = file_get_contents($fixtureXmlFile);
+
+        $file = $this->createMock(\Magento\Framework\Filesystem\File\Read::class);
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: Import class

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+            });
+
+        $this->readFactory = $this->createMock(ReadFactory::class);
+        $this->readFactory->expects($this->once())
+            ->method('create')
+            ->withAnyParameters()
+            ->willReturn($currentDirRead);
+
+        $this->fileWriteFactory = $this->createMock(WriteFactory::class);
+        $this->domFactory = new DomDocumentFactory();
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: Use ObjectManager helper `\Magento\Framework\TestFramework\Unit\Helper\ObjectManager`

File: app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+     */
+    protected $readFactory;
+
+    /**
+     * @var Magento\Framework\Filesystem\File\WriteFactory
+     */
+    protected $fileWriteFactory;
+
+    /**
+     * @var Magento\Framework\DomDocument\DomDocumentFactory
```
Comment (lbajsarowicz) [inline app/code/Magento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php]: Import class.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
We still need you to do some improvements.

**Comment by lbajsarowicz** [comment]
Static Analysis reports missing Copyright in new files and line width exceeded.
Address these issues and we're ready to approve : )

**Comment by lbajsarowicz** [comment]
@manuelcanepa Last thing to be fixed:
```
FILE: ...gento/Developer/Test/Unit/Model/XmlCatalog/Format/VsCodeTest.php
----------------------------------------------------------------------
FOUND 0 ERRORS AND 2 WARNINGS AFFECTING 2 LINES
----------------------------------------------------------------------
 271 | WARNING | [ ] Line exceeds 120 characters; contains 155
     |         |     characters
 277 | WARNING | [x] The closing brace for the class must go on the
     |         |     next line after the body
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

**Review by lbajsarowicz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-6554](https://jira.corp.magento.com/browse/ENGCOM-6554) has been created to process this Pull Request

**Review by Nazar65 — APPROVED** [review]

**Comment by engcom-Alfa** [comment]
@lbajsarowicz Could you review the last commits? Thanks!

**Review by lbajsarowicz — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-6554](https://jira.corp.magento.com/browse/ENGCOM-6554) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

File: app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+     */
+    private function initEmptyFile(\DOMDocument $dom): \DOMElement
+    {
+        $copyrigthComment = $dom->createComment('
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+');
+        $dom->appendChild($copyrigthComment);
```
Comment (ihor-sviziev) [inline app/code/Magento/Developer/Model/XmlCatalog/Format/VsCode.php]: here is typo.
```suggestion
        $dom->appendChild($copyrightComment);
```

**Pull request merged**